theano/tensor/opt.py

```diff
@@ -2669,6 +2673,107 @@ def local_adv_sub1_adv_inc_sub1(node):
     return [T.cast(y, node.outputs[0].dtype)]
 
 
+@register_specialize
+@register_stabilize
+@register_canonicalize
+@gof.local_optimizer([IncSubtensor,
+                      AdvancedIncSubtensor,
+                      AdvancedIncSubtensor1])
+def local_useless_inc_subtensor_alloc(node):
+    """
+    Replaces an [Advanced]IncSubtensor[1], whose increment is an `alloc` of
+    a fully or partially broadcastable variable, by one that skips the
+    intermediate `alloc` where possible.
+    """
+    if isinstance(node.op, (IncSubtensor,
+                            AdvancedIncSubtensor,
+                            AdvancedIncSubtensor1)):
+        x = node.inputs[0]
+        y = node.inputs[1]
+        i = node.inputs[2:]
+
+        if y.owner is not None and isinstance(y.owner.op, T.Alloc):
+            # `z` is the input of the Alloc op, i.e. T.alloc(z, <shape>)
+            z = y.owner.inputs[0]
+
+            try:
+                shape_feature = node.fgraph.shape_feature
+            except AttributeError:
+                # The shape feature may not be available in some mode, but we
+                # need it for this optimization, so don't continue.
+                return False
+
+            shape_of = shape_feature.shape_of
+            same_shape = shape_feature.same_shape
+
+            # Get the subtensor of `x` indexed by `i` in order to compare
+            # shapes later.
+            if isinstance(node.op, IncSubtensor):
+                xi = Subtensor(node.op.idx_list)(x, *i)
+            elif isinstance(node.op, AdvancedIncSubtensor):
+                xi = advanced_subtensor(x, *i)
+            elif isinstance(node.op, AdvancedIncSubtensor1):
+                xi = advanced_subtensor1(x, *i)
+            else:
+                raise Exception('Should never happen!')
+
+            reason = 'local_useless_incsubtensor_alloc'
+
+            # Add `xi` to the shape feature `fgraph`. This is important for
+            # shape inference later because the variable must be part of the
+            # function graph in order to call `same_shape` on it.
+            if xi not in shape_of:
```

Comment:
If you do:
x=theano.tensor.matrix()
y=x+1
z=x+1

Then y and z are 2 different variable. Theano will merge them during
compilation. In your case, even if you are during compilation, you just
created a new node that haven't had a chance to be merged yet. So it can't
be in shape_of now.

Also, I just check the code of on_import, if it is already in shape_of, it
do nothing (just assert some stuff).

On Wed, Apr 22, 2015 at 10:08 AM, sisp <notifications@github.com> wrote:

> In theano/tensor/opt.py
> <https://github.com/Theano/Theano/pull/2521#discussion_r28875053>:
>
> > +            # shapes later.
> > +            if isinstance(node.op, IncSubtensor):
> > +                xi = Subtensor(node.op.idx_list)(x, *i)
> > +            elif isinstance(node.op, AdvancedIncSubtensor):
> > +                xi = advanced_subtensor(x, *i)
> > +            elif isinstance(node.op, AdvancedIncSubtensor1):
> > +                xi = advanced_subtensor1(x, *i)
> > +            else:
> > +                raise Exception('Should never happen!')
> > +
> > +            reason = 'local_useless_incsubtensor_alloc'
> > +
> > +            # Add `xi` to the shape feature `fgraph`. This is important for
> > +            # shape inference later because the variable must be part of the
> > +            # function graph in order to call `same_shape` on it.
> > +            if xi not in shape_of:
>
> Isn't it possible that xi was already created in the graph somewhere
> else? Thus, it is already in shape_of and we just created the same
> variable again here? Or would they be considered different variables at
> this point?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2521/files#r28875053>.
>